### PR TITLE
Info relation refactor

### PR DESCRIPTION
**Pull request by yuli6313**
refactor info relation in CompiledSouffle and SynthesisrRelation

### CODE REVIEW

File: src/CompiledSouffle.h [hunk 1]
```diff
@@ -234,4 +234,97 @@ class t_nullaries {
...
+        }
+    };
+    iterator begin() const {
+        return iterator(data.data());
+    }
+    iterator end() const {
+        return iterator(data.data() + data.size());
+    }
+    void insert(const t_tuple& t) {
+        insert_lock.lock();
```
Comment (b-scholz) [inline src/CompiledSouffle.h]: Please check whether it is contained in the vector first before inserting it into the vector.

File: src/CompiledSouffle.h [hunk 2]
```diff
@@ -234,4 +234,97 @@ class t_nullaries {
...
+        data.push_back(t);
+        insert_lock.unlock();
+    }
+    void insert(const t_tuple& t, context& /* ctxt */) {
+        insert_lock.lock();
```
Comment (b-scholz) [inline src/CompiledSouffle.h]: Please reuse the previous insert() method.

File: src/CompiledSouffle.h [hunk 3]
```diff
@@ -234,4 +234,97 @@ class t_nullaries {
...
+        data.push_back(t);
+        insert_lock.unlock();
+    }
+    void insert(const RamDomain* ramDomain) {
+        insert_lock.lock();
```
Comment (b-scholz) [inline src/CompiledSouffle.h]: Please reuse the previous insert() method rather than replicating code.

File: src/CompiledSouffle.h [hunk 4]
```diff
@@ -234,4 +234,97 @@ class t_nullaries {
...
+    bool contains(const t_tuple& t) const {
+        for (const auto& o : data) {
+            if (t == o) {
+                return true;
+            }
+        }
+        return false;
+    }
+    bool contains(const t_tuple& t, context& /* ctxt */) const {
+        for (const auto& o : data) {
```
Comment (b-scholz) [inline src/CompiledSouffle.h]: Reuse the previous method contains() instead of replicating it.

**Comment by b-scholz** [comment]
You need to run ```clang-format -i *.cpp *.h``` in your directory.

**Comment by b-scholz** [comment]
You still have a compile issue:
```
In file included from InterpreterRelation.cpp:17:
InterpreterRelation.h:272:2: error: expected ';' after class definition
  272 | }
      |  ^
      |  ;
```

File: src/Synthesiser.cpp [hunk 5]
```diff
@@ -1683,9 +1683,9 @@ void Synthesiser::generateCode(std::ostream& os, const std::string& id, bool& wi
 
     // synthesise data-structures for relations
     for (auto rel : prog.getRelations()) {
-        const std::string& datalogName = rel->getName();
+        // const std::string& datalogName = rel->getName();
```
Comment (mmcgr) [inline src/Synthesiser.cpp]: Delete

File: src/SynthesiserRelation.cpp [hunk 6]
```diff
@@ -50,6 +52,24 @@ std::unique_ptr<SynthesiserRelation> SynthesiserRelation::getSynthesiserRelation
     return std::unique_ptr<SynthesiserRelation>(rel);
 }
 
+// -------- Info Relation --------
+
+/** Generate index set for a nullary relation, which should be empty */
```
Comment (mmcgr) [inline src/SynthesiserRelation.cpp]: This is not a nullary. Same below.

File: src/CompiledSouffle.h [hunk 7]
```diff
@@ -234,4 +234,92 @@ class t_nullaries {
...
+    void insert(const RamDomain* ramDomain) {
+        insert_lock.lock();
+        t_tuple t;
+        for (size_t i = 0; i < Arity; ++i) {
+            t.data[i] = ramDomain[i];
+        }
+        data.push_back(t);
+        insert_lock.unlock();
+    }
+    bool insert() {
```
Comment (b-scholz) [inline src/CompiledSouffle.h]: I don't think we need this method. This is a legacy of the nullary class?

Comment (yuli6313) [reply src/CompiledSouffle.h]: It is legacy of nullary relation. I ll delete it

**Comment by b-scholz** [comment]
This is related to issue #1202

**Pull request merged**